nnstreamer/tensor_trainer/tensor_trainer_nntrainer.hh

```diff
@@ -0,0 +1,129 @@
+ * @brief  NNStreamer tensor_trainer subplugin header
+ * @see    http://github.com/nnstreamer/nnstreamer
+ * @author Hyunil Park <hyunil46.park@samsung.com>
+ * @bug    No known bugs except for NYI items
+ */
+
+#include <model.h>
+#include <nnstreamer_plugin_api.h>
+#include <nnstreamer_plugin_api_trainer.h>
+#include <vector>
+
+namespace NNTrainer {
+
+/**
+ * @brief Manage multiple inputs and labels data
+ */
+struct TensorData {
+  std::vector<char *> inputs;
+  std::vector<char *> labels;
+};
+
+class InputTensorsInfo;
+
+/**
+ * @brief NNTrainer interface for nnstreamer trainer subplugin
+ */
+class NNTrainerTrain {
+public:
+  /**
+   * @brief Construct a new NNTrainerTrain object
+   * @param prop tensor trainer subplugin properties
+   * @param _model_config model configuration file path
+   */
+  NNTrainerTrain(const GstTensorTrainerProperties *prop,
+                 const std::string &_model_config);
+
+  /**
+   * @brief Destroy the NNTrainerTrain object
+   */
+  ~NNTrainerTrain() = default;
+
+  /**
+   * @brief Create model
+   */
+  void createModel();
+
+  /**
+   * @brief Train model
+   */
+  void trainModel();
+
+  /**
+   * @brief Create dataset
+   */
+  void createDataset();
+
+  /**
+   * @brief Get NNStreamer tensor_trainer properties
+   * @param prop Tensor trainer subplugin properties
+   */
+  void getNNStreamerProperties(const GstTensorTrainerProperties *prop);
+
+  /**
+   * @brief Manage sample data
+   */
+  std::unique_ptr<NNTrainer::InputTensorsInfo> train_data, valid_data;
+  /**
+   * @brief Nntrainer dataset
+   */
+  std::shared_ptr<ml::train::Dataset> dataset_train, dataset_valid;
+  float training_loss, validation_loss;
+
+  unsigned int tensors_inputsize[NNS_TENSOR_SIZE_LIMIT];
+  unsigned int num_tensors;
+  unsigned int num_inputs;
+  unsigned int num_labels;
+  unsigned int num_train_samples;
+  unsigned int num_valid_samples;
+  std::string model_config;
+  std::string model_save_path;
+
+  GCond *train_complete_cond;
+
+private:
+  std::unique_ptr<ml::train::Model> model;
+};
+
+/**
+ * @brief Manage input tensors data and information
+ */
+class InputTensorsInfo {
+public:
+  /**
+   * @brief Construct a new InputTensorsInfo object
+   * @param _num_samples number of samples
+   * @param _num_inputs number of inputs
+   * @param _num_labels number of labels
+   * @param _tensors_inputsize[] input tensors size
+   */
+  InputTensorsInfo(unsigned int _num_samples, unsigned int _num_inputs,
+                   unsigned int num_labels, unsigned int _tensors_inputsize[]);
+
+  /**
+   * @brief Destroy the InputTensorsInfo object
+   */
+  ~InputTensorsInfo();
+
+  bool is_mutex_locked;
+  int64_t push_count;
+  int64_t pop_count;
+  int64_t input_size[NNS_TENSOR_SIZE_LIMIT]; // feature size * data type
+  int64_t label_size[NNS_TENSOR_SIZE_LIMIT];
+  int64_t num_samples;
+  int64_t num_inputs;
+  int64_t num_labels;
+
+  std::vector<TensorData> tensor_data;
+  pthread_mutex_t mutex;
```

Comment:
https://en.cppreference.com/w/cpp/thread/mutex/lock

How about using CPP's lock methods?

> Notes
lock() is usually not called directly: [std::unique_lock](https://en.cppreference.com/w/cpp/thread/unique_lock), [std::scoped_lock](https://en.cppreference.com/w/cpp/thread/scoped_lock), and [std::lock_guard](https://en.cppreference.com/w/cpp/thread/lock_guard) are used to manage exclusive locking.

Reply:
Thank you. I will change.

Reply:
Done ([0904ed7](https://github.com/nnstreamer/nntrainer/pull/2059/commits/0904ed77922ad0b8263acb2bdcfc539047a3182e)), Thank you.